### DISCUSSION TITLE
100/ naming

### DISCUSSION THREAD

**Discussion by jkotlinski** [jkotlinski/durexforth#543] [General]
I don't anymore particularly like the naming of `100/` (which shifts top of stack right by 8, inserting 0)

- 100 no longer makes sense, because durexForth now uses DECIMAL base by default
- / does not make total sense, because it implies arithmetic right shift.

Any ideas? Are there any known precedents?

Right now, I'm leaning towards 256/...

**Comment by ekipan** [comment]
I was actually thinking the same while writing my own multiply and divide words! I usually name them with decimal. It's just a style thing so it doesn't actually matter but my preference is for `256/`. I guess `$100/` or `#256/` would be clearer but with 40 columns for source code space is at a premium.

**Reply by ekipan** [reply]
I'm poking at a block editor proof of concept. It uses these words:
```forth
: 32/  2/ 2/ 2/ 2/ 2/ ;
: 40*  dup 2* 2* + 2* 2* 2* ;
: 1024* [ lsb lda,x asl,a asl,a
  msb sta,x 0 lda,# lsb sta,x ] ;
```

**Comment by Whammo** [comment]
It's strange, that an operation so ubiquitous be not standardized. Perhaps, simply documenting which way of going about it is fastest, and then leaving to the user.
 Though, it did occur to me that a series of operations could be created to treat the top of stack as if it were a bit stack.

**Reply by ekipan** [reply]
Something like this?
```forth
: popbit  dup 1 and swap 2/ ;
: pushbit  2* or ;
T{ %101 popbit -> 1 %10 }T
T{ 1 %10 pushbit -> %101 }T
```

**Reply by Whammo** [reply]
Careful, or we could introduce ```bitrot```

😄

**Comment by Whammo** [comment]
So many ways to accomplish the operation.

code 256/
msb lda, lsb sta, 0 lda,# msb sta, 
;code

Edit: oh.

**Reply by ekipan** [reply]
The [builtin one](https://github.com/jkotlinski/durexforth/blob/485e245f4c56b6894ac3e8e762b1ba7efad85b5f/asm/core.asm#L415-L420) is only four instructions.

Actually that stack effect comment is wrong! Should be `( u1 -- u2 )` @jkotlinski 

Edit: @Whammo it won't work because you have to index by the X register. You could do this:
```forth
: 256/ [ 0 ldy,# msb lda,x ] pushya nip ;
```
But that's just wasting a bunch of loads, stores, and an inx and dex pair.

**Comment by jkotlinski** [comment]
How about HIBYTE?

**Reply by ekipan** [reply]
~I guess `100/` also isn't ideal since unlike `/` it doesn't preserve sign.~ (Yeah you did mention this already in your post.) I used these names in a program I wrote for a different forth, though I don't think my naming convention> for >pure words is established or anything.
```forth
: >lo  $ff and ;
: >hi  8 rshift >lo ;
: lohi>  8 lshift or ;
```

**Reply by jkotlinski** [reply]
Not bad! I researched a bit, what different 8-bit assemblers use. `>` for high-byte seems very common. Then `HIGH`, `HIBYTE`, `HIGHBYTE` as spell-it-out options. `>hi` is a bit of a mix.

I like `hibyte` because it makes sense when read aloud. Also looks a bit similar to `rshift`, which is the closest related word.

**Reply by jkotlinski** [reply]
In present-day 8-bit assemblers and compilers, `HIGH` seems to be the most common choice.

**Reply by ekipan** [reply]
```forth
code >lohi ( u1 -- u2 u3 )
msb lda,x dex, lsb sta,x 
0 lda,# msb sta,x msb 1+ sta,x
;code

code lohi> ( u1 u2 -- u3 )
lsb lda,x inx, msb sta,x
;code
```
Re: your [other comment](https://github.com/jkotlinski/durexforth/discussions/543#discussioncomment-5654572) as well. This is with `>lo` and `>hi` coupled together. If you need both bytes I guess this is fewer instructions.

**Reply by jkotlinski** [reply]
`>lohi` could be named `split` or possibly `bytesplit`. Seems like a very useful word.

**Reply by jkotlinski** [reply]
Yeah... I think `split` fits perfectly. Best option so far...

**Reply by jkotlinski** [reply]
I really love `split` :) Super useful.

**Comment by jkotlinski** [comment]
Another idea: Replace `100/` with

    code byteswap
    lda msb,x ldy lsb,x
    sty msb,x sta lsb,x
    ;code

**Reply by jkotlinski** [reply]
Good:

- obvious naming (or?)
- is more generally useful than `100/`
- usually works as drop-in replacement for `100/`

Bad:

- is not always drop-in replacement for `100/`

**Comment by Whammo** [comment]
```c!``` and the like ignores the high byte.
Perhaps, ``` c>!``` and the like ignores the low.

**Reply by Whammo** [reply]
```forth
: 256/ w c>! w c@ ;
```

**Comment by jkotlinski** [comment]
How it turned out: https://github.com/jkotlinski/durexforth/pull/546
